### PR TITLE
DM: update optimistic shard ddl

### PR DESCRIPTION
**Pull request by GMHDBJD**
### First-time contributors' checklist 

- [x] I've signed [**Contributor License Agreement**](https://cla-assistant.io/pingcap/docs-cn) that's required for repo owners to accept my contribution.

### Which TiDB version(s) do your changes apply to? (Required)

**Tips for choosing the affected version(s):**

By default, **CHOOSE MASTER ONLY** so your changes will be applied to the next TiDB major or minor releases. If your PR involves a product feature behavior change or a compatibility change, **CHOOSE THE AFFECTED RELEASE BRANCH(ES) AND MASTER**.

For details, see [tips for choosing the affected versions (in Chinese)](https://github.com/pingcap/docs-cn/blob/master/CONTRIBUTING.md#版本选择指南).

- [x] master (the latest development version)
- [x] v6.1 (TiDB 6.1 versions)
- [ ] v6.0 (TiDB 6.0 versions)
- [ ] v5.4 (TiDB 5.4 versions)
- [ ] v5.3 (TiDB 5.3 versions)
- [ ] v5.2 (TiDB 5.2 versions)
- [ ] v5.1 (TiDB 5.1 versions)
- [ ] v5.0 (TiDB 5.0 versions)
- [ ] v4.0 (TiDB 4.0 versions)

### What is the related PR or file link(s)?

- This PR is translated from:
- Other reference link(s): https://github.com/pingcap/tiflow/pull/5198

### Do your changes match any of the following descriptions?

- [ ] Delete files
- [ ] Change aliases
- [ ] Need modification after applied to another branch 
- [ ] Might cause conflicts after applied to another branch

### CODE REVIEW

File: dm/feature-shard-merge-optimistic.md [hunk 1]
```diff
@@ -32,18 +32,26 @@ DM 支持在线上执行分库分表的 DDL 语句（通称 Sharding DDL），
     - 例如：在分表执行过 `ADD COLUMN A INT; DROP COLUMN A; ADD COLUMN A FLOAT;`，在其他分表直接执行 `ADD COLUMN A FLOAT` 即可，不需要三条 DDL 都执行一遍。
 - 执行 DDL 时要注意观察 DM 迁移状态。当迁移报错时，需要判断这个批次的 DDL 是否会造成数据不一致。
 
-“乐观协调”模式暂不支持以下语句：
+“乐观协调”模式时，上游在执行**不兼容**的 DDL 时，需要**保证该 DDL 在各分表中按相同的顺序**执行。**不兼容**的 DDL 是指改变列名，列属性，或列默认值的 DDL。如：
```
Comment (ran-huang) [inline dm/feature-shard-merge-optimistic.md:35]:
```suggestion
“乐观协调”模式下，上游在执行**不兼容**的 DDL 时，需要**保证该 DDL 在各分表中按相同的顺序**执行。**不兼容**的 DDL 是指改变列名、列属性或列默认值的 DDL。如：
```

File: dm/feature-shard-merge-optimistic.md [hunk 2]
```diff
@@ -32,18 +32,26 @@ DM 支持在线上执行分库分表的 DDL 语句（通称 Sharding DDL），
...
-- `ALTER TABLE table_name ADD COLUMN column_name datetime DEFAULT NOW()`（增加的列默认值不固定）。
-- `ALTER TABLE table_name ADD COLUMN col1 INT, DROP COLUMN col2`（在一个 DDL 语句中同时包含 `ADD COLUMN` 与 `DROP COLUMN`）。
+- `ALTER TABLE table_name MODIFY COLUMN column_name VARCHAR(20)`（修改列的类型）。
 - `ALTER TABLE table_name RENAME COLUMN column_1 TO column_2;`（重命名列）。
+- `ALTER TABLE table_name ADD COLUMN column_1 NOT NULL;`（增加没有默认值且非空的列）。
 - `ALTER TABLE table_name RENAME INDEX index_1 TO index_2;`（重命名索引）。
 
+各分表在执行以上 DDL 时，若顺序不同，将导致同步中断，如：
+
+- 分表 1 先重名列，再修改列类型
```
Comment (ran-huang) [inline dm/feature-shard-merge-optimistic.md:44]:
```suggestion
- 分表 1 先重命名列，再修改列类型
```

File: dm/feature-shard-merge-optimistic.md [hunk 3]
```diff
@@ -32,18 +32,26 @@ DM 支持在线上执行分库分表的 DDL 语句（通称 Sharding DDL），
...
+    1. `ALTER TABLE table_name RENAME COLUMN column_1 TO column_2;`（重命名列）。
+    2. `ALTER TABLE table_name MODIFY COLUMN column_3 VARCHAR(20);`（修改列类型）。
+- 分表 2 先修改列类型，再重命名列
+    1. `ALTER TABLE table_name MODIFY COLUMN column_3 VARCHAR(20)`（修改列的类型）。
```
Comment (ran-huang) [inline dm/feature-shard-merge-optimistic.md:48]:
```suggestion
    1. `ALTER TABLE table_name MODIFY COLUMN column_3 VARCHAR(20)`（修改列类型）。
```

**Review by ran-huang — COMMENTED** [review]
Rest LGTM

File: dm/feature-shard-merge-optimistic.md (hunk 1 above)
Comment (sunzhaoyang) [inline dm/feature-shard-merge-optimistic.md:35]: ”不兼容“ 这个词感觉可以换一下或者再解释一下。可能会有人看到不兼容以为是无法执行，但其实只是有限制条件而已。

Comment (ran-huang) [reply dm/feature-shard-merge-optimistic.md:35]: 加一句解释吧：“不兼容指.....（DM 行为）“ @GMHDBJD

Comment (GMHDBJD) [reply dm/feature-shard-merge-optimistic.md]: > ”不兼容“ 这个词感觉可以换一下或者再解释一下

File: dm/feature-shard-merge-optimistic.md [hunk 4]
```diff
@@ -32,18 +32,26 @@ DM 支持在线上执行分库分表的 DDL 语句（通称 Sharding DDL），
     - 例如：在分表执行过 `ADD COLUMN A INT; DROP COLUMN A; ADD COLUMN A FLOAT;`，在其他分表直接执行 `ADD COLUMN A FLOAT` 即可，不需要三条 DDL 都执行一遍。
 - 执行 DDL 时要注意观察 DM 迁移状态。当迁移报错时，需要判断这个批次的 DDL 是否会造成数据不一致。
 
-“乐观协调”模式暂不支持以下语句：
+“乐观协调”模式下，上游执行的大部分 DDL 无需特别关注，将自动同步，本文使用“一类 DDL” 指代。同时，还有一部分改变列名、列属性或列默认值的 DDL，称为“二类 DDL”，上游执行时需要特别注意： 上游在执行时必须**保证该 DDL 在各分表中按相同的顺序**执行。
```
Comment (ran-huang) [inline dm/feature-shard-merge-optimistic.md:35]:
```suggestion
“乐观协调”模式下，上游执行的大部分 DDL 无需特别关注，将自动同步，本文使用“一类 DDL” 指代。同时，还有一部分改变列名、列属性或列默认值的 DDL，称为“二类 DDL”，上游执行时要注意必须**保证该 DDL 在各分表中按相同的顺序**执行。
```

File: dm/feature-shard-merge-optimistic.md [hunk 5]
```diff
@@ -32,18 +32,26 @@ DM 支持在线上执行分库分表的 DDL 语句（通称 Sharding DDL），
...
 
-- `ALTER TABLE table_name ADD COLUMN column_name datatype NOT NULL`（添加无默认值的 not null 的列）。
-- `ALTER TABLE table_name ADD COLUMN column_name datetime DEFAULT NOW()`（增加的列默认值不固定）。
-- `ALTER TABLE table_name ADD COLUMN col1 INT, DROP COLUMN col2`（在一个 DDL 语句中同时包含 `ADD COLUMN` 与 `DROP COLUMN`）。
+“二类 DDL” 举例如下：
+
+- `ALTER TABLE table_name MODIFY COLUMN column_name VARCHAR(20)`（修改列的类型）。
 - `ALTER TABLE table_name RENAME COLUMN column_1 TO column_2;`（重命名列）。
+- `ALTER TABLE table_name ADD COLUMN column_1 NOT NULL;`（增加没有默认值且非空的列）。
 - `ALTER TABLE table_name RENAME INDEX index_1 TO index_2;`（重命名索引）。
```
Comment (ran-huang) [inline dm/feature-shard-merge-optimistic.md:42]:
```suggestion
- 修改列的类型：`ALTER TABLE table_name MODIFY COLUMN column_name VARCHAR(20)`。
- 重命名列：`ALTER TABLE table_name RENAME COLUMN column_1 TO column_2;`。
- 增加没有默认值且非空的列：`ALTER TABLE table_name ADD COLUMN column_1 NOT NULL;`。
- 重命名索引：`ALTER TABLE table_name RENAME INDEX index_1 TO index_2;`。
```

File: dm/feature-shard-merge-optimistic.md [hunk 6]
```diff
@@ -32,18 +32,26 @@ DM 支持在线上执行分库分表的 DDL 语句（通称 Sharding DDL），
...
 
+各分表在执行以上 DDL 时，若顺序不同将导致同步中断，例如下述场景：
+
+- 分表 1 先重命名列，再修改列类型
+    1. `ALTER TABLE table_name RENAME COLUMN column_1 TO column_2;`（重命名列）。
+    2. `ALTER TABLE table_name MODIFY COLUMN column_3 VARCHAR(20);`（修改列类型）。
```
Comment (ran-huang) [inline dm/feature-shard-merge-optimistic.md:48]:
```suggestion
    1. 重命名列：`ALTER TABLE table_name RENAME COLUMN column_1 TO column_2;`。
    2. 修改列类型：`ALTER TABLE table_name MODIFY COLUMN column_3 VARCHAR(20);`。
```

File: dm/feature-shard-merge-optimistic.md [hunk 7]
```diff
@@ -32,18 +32,26 @@ DM 支持在线上执行分库分表的 DDL 语句（通称 Sharding DDL），
...
+- 分表 2 先修改列类型，再重命名列
+    1. `ALTER TABLE table_name MODIFY COLUMN column_3 VARCHAR(20)`（修改列类型）。
+    2. `ALTER TABLE table_name RENAME COLUMN column_1 TO column_2;`（重命名列）。
```
Comment (ran-huang) [inline dm/feature-shard-merge-optimistic.md:51]:
```suggestion
    1. 修改列类型：`ALTER TABLE table_name MODIFY COLUMN column_3 VARCHAR(20)`。
    2. 重命名列：`ALTER TABLE table_name RENAME COLUMN column_1 TO column_2;`。
```

**Review by ran-huang — APPROVED** [review]

**Review by sunzhaoyang — APPROVED** [review]
LGTM

**Pull request merged**